modules/node_groups/node_groups.tf

```
@@ -0,0 +1,49 @@
+resource "aws_eks_node_group" "workers" {
+  for_each = random_pet.node_groups
+
+  node_group_name = join("-", [var.cluster_name, each.key, each.value.id])
+
+  cluster_name  = var.cluster_name
+  node_role_arn = each.value.keepers.node_role_arn
+  subnet_ids    = split("|", each.value.keepers.subnet_ids)
+
+  scaling_config {
+    desired_size = lookup(local.node_groups_expanded[each.key], "desired_capacity", var.workers_group_defaults["asg_desired_capacity"])
+    max_size     = lookup(local.node_groups_expanded[each.key], "max_capacity", var.workers_group_defaults["asg_max_size"])
+    min_size     = lookup(local.node_groups_expanded[each.key], "min_capacity", var.workers_group_defaults["asg_min_size"])
+  }
+
+  ami_type        = lookup(each.value.keepers, "ami_type", null)
+  disk_size       = lookup(each.value.keepers, "disk_size", null)
+  instance_types  = [each.value.keepers.instance_type]
+  release_version = lookup(local.node_groups_expanded[each.key], "ami_release_version", null)
+
+  dynamic "remote_access" {
+    for_each = each.value.keepers.ec2_ssh_key != "" ? [{
+      ec2_ssh_key               = each.value.keepers.ec2_ssh_key
+      source_security_group_ids = compact(split("|", each.value.keepers.source_security_group_ids))
+    }] : []
+
+    content {
+      ec2_ssh_key               = remote_access.value["ec2_ssh_key"]
+      source_security_group_ids = remote_access.value["source_security_group_ids"]
+    }
+  }
+
+  version = var.cluster_version
```

Allowing a user to set a node group's value means that they can incrementally replace their node groups as part of an upgrade process rather than having every single node group replaced simultaneously, which is what this currently does. I strongly recommend that instead we do something along the lines of the following (some tweaking may be required):
```suggestion
  version = lookup(var.node_groups[each.value], "version", null)
```

Note that we're able to use `null` since [the default is for node groups to match the EKS cluster version](https://www.terraform.io/docs/providers/aws/r/eks_node_group.html#version):
> version – (Optional) Kubernetes version. Defaults to EKS Cluster Kubernetes version. Terraform will only perform drift detection if a configuration value is provided.

---

I agree. This is existing logic. I'll change it to be user configurable.